Fix deadlocks resulting from the post-kmem rework merge

dweeezil opened the pull request:
Put the following functions under PF_FSTRANS to prevent deadlocks
due to re-entering the filesystem during reclaim:

	dbuf_hold()
	dmu_prefetch()
	sa_buf_hold()

Testing this one now:
	dmu_zfetch_fetch() - z_hold_mtx[] via zget and zf_rwlock via dmu_zfetch

dweeezil commented:
This is intended to fix the issues outlined in #3183.  It's a WIP branch but since it appears others have been grabbing patches from it, I think it's time to get it posted as a pull request.

The "testing this one now" change to ```dmu_zfetch_zfetch()``` was added late yesterday to fix a deadlock I recently outlined but it hasn't been tested heavily yet.  There's a very good chance I pushed the fstrans marking too far down the call chain so further evaluation is necessary.

dweeezil commented:
Unfortunately, it seems ```z_hold_mtx[i]``` strikes again and does so in a manner that I feared could happen.  It's possible for one process to hold entry X via the ```zfs_zget()``` path and try to acquire entry Y via ```zfs_zinactive()``` while another process holding entry Y via ```zfs_zinactive()``` enters reclaim in the ```zfs_zget->zfs_znode_alloc-``` path and winds up in ```zfs_zinactive()``` trying to acquire X.

I'm going to try the more heavy-handed approach of putting ```z_hold_mtx[]``` under MUTEX_FSTRANS, but I suspect the plan was to use that feature as a last resort.  I had to put the system under the most extreme memory pressure possible in order to catch this one; the moral equivalent of a "reclaim torture test" mentioned in another issue.

tuxoko commented:
@dweeezil 
You HAVE TO put z_hold_mtx under FSTRANS. Otherwise, you'll have to add spl_fstrans_mark to every possible memory allocation under z_hold_mtx. The reason is just as you described, direct reclaim will cause reentry on z_hold_mtx, and it will cause deadlock even if the reentry is on different z_hold_mtx entry, because there could be multiple threads doing reclaim.

dweeezil commented:
@tuxoko I'll be running tests with z_hold_mtx under MUTEX_FSTRANS shortly.  I expect it will fix this current case of deadlocks.  There are clearly too many code paths to fix, individually, otherwise.

tuxoko commented:
@dweeezil @behlendorf 
By the way, I'm not sure why nobody commented on this: https://github.com/zfsonlinux/zfs/issues/3183#issuecomment-85425528 ...
As far as I can tell, after I added MUTEX_FSTRANS to l2arc_buflist_mtx, the order was already violated between hash_lock in arc_release().

If we are going to keep the MUTEX_FSTRANS thing as it is, we need to be very careful of the lock/unlock sequence of such mutexes. spl_fstrans_mark would also have effect in this. But personally, unless we could fix the MUTEX_FSTRANS thing, I think we should stick to spl_fstrans_mark, because it is more explicit.

dweeezil commented:
@tuxoko Sorry, I had been meaning to read the comment you referenced in #3183 more closely but since I had not been leaning on MUTEX_FSTRANS, I didn't bother.  Your comment is, of course, correct and is certainly a potential problem with z_hold_mtx in which the entries are aliased among potentially many unrelated objects.  Using MUTEX_FSTRANS as a fix for z_hold_mtx isn't going to work as it stands now.

tuxoko commented:
@dweeezil 
As far as I can tell, z_hold_mtx is pretty well behaved. It always acquires and releases in the same functon. So it should be pretty straight forward to use spl_fstrans_mark on them.

dweeezil commented:
My latest commit simply marks/unmarks where the z_hold_mtx array is used.  This seems to cure all the cases of its related deadlocks I've been able to produce so far.  I've got one small related tweak I'll push later today but I think this is a good start.

Unfortunately, I can still deadlock my test system with the stress test.  I've not analyzed the stack traces and lock debugging fully yet but it looks like it may be related to the kernel's per-inode ```i_mutex``` lock.  I'm concerned this one may be more an interaction between the VFS and ZFS issue rather than strictly a ZFS issue.  It looks more like a simple deadlock rather than a lock inversion problem.  My specific concern is that something in the manner in which directories are mapped to objects and/or their parentage is managed under ZFS may be incompatible with the assumptions made by the VFS insofar as managing ```i_mutex``` is concerned.  I'll know more after analyzing the stacks I've captured.

dweeezil commented:
I've pushed another version of this with the comments cleaned up and one small superfluous change reverted.  Further testing shows that the deadlocks I was still able to create are somewhat expected on systems with very little memory, no swap and on which tmpfs is being used.  After removing all uses of tmpfs, I was no longer able to create the same deadlock.  That said, I'm going to continue stress testing over the next day or so.

dweeezil commented:
Here's an update after chasing deadlocks all day.  Most of the last class of deadlocks I was seeing were due to direct reclaim during the various dmu_tx_hold family of functions.  Unfortunately, after locking all of those paths down, I'm now able generate deadlocks involving the ```metaslab_preload()...arc_buf_alloc()``` path.  At this rate, the entire system will be reduced to non-sleeping allocations just as it mostly was before the kmem-rework merge.  Furthermore, all these semi-gratuitous changes cause even more divergence with upstream code than KM_PUSHPAGE did in the past.  I'm going to keep plugging away at it for awhile but a little voice in the back of my head is telling me we need a different approach.

dweeezil commented:
I've backed off of all the dmu_tx_hold lockdowns.  Instead, as @tuxoko rightly pointed out, memory allocations when db_mtx is held can cause deadlocks.  The latest commit puts it under MUTEX_FSTRANS but since I've got a good set of stack traces involving this deadlock, it might be possible to do this in a more granular manner with ```spl_fstrans_mark()```.

With this patch in place, however, I can still lock up my test system in which it spins trying to free memory in the SPL slab shrinker.  I'm going to get this rebased on current master code now in order to get the latest ARC changes (2cbb06b in particular).

kernelOfTruth commented:
hm, several git fetch/pull fails due to github being DDOS

and

http://buildbot.zfsonlinux.org/builders/centos-7.0-x86_64-builder/builds/1764/steps/shell_17
```
[ 5959.028703] WARNING: at fs/xfs/xfs_aops.c:968 xfs_vm_writepage+0x5ab/0x5c0 [xfs]()
[ 5959.028704] Modules linked in: zfs(POF) zunicode(POF) zavl(POF) zcommon(POF) znvpair(POF) spl(OF) sd_mod crct10dif_generic crc_t10dif crct10dif_common ext4 mbcache jbd2 loop zlib_deflate ip6t_rpfilter ip6t_REJECT ipt_REJECT xt_conntrack ebtable_nat ebtable_broute bridge stp llc ebtable_filter ebtables ip6table_nat nf_conntrack_ipv6 nf_defrag_ipv6 nf_nat_ipv6 ip6table_mangle ip6table_security ip6table_raw ip6table_filter ip6_tables iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat nf_conntrack iptable_mangle iptable_security iptable_raw iptable_filter ip_tables sg crc32c_intel aesni_intel lrw gf128mul glue_helper ablk_helper cryptd virtio_console serio_raw pcspkr virtio_balloon i2c_piix4 mperf xfs libcrc32c sr_mod cdrom ata_generic pata_acpi virtio_blk virtio_scsi qxl virtio_net drm_kms_helper
[ 5959.028743]  ttm drm ata_piix libata virtio_pci virtio_ring virtio i2c_core floppy dm_mirror dm_region_hash dm_log dm_mod [last unloaded: scsi_debug]
[ 5959.028752] CPU: 2 PID: 10982 Comm: kworker/u8:2 Tainted: PF       W  O--------------   3.10.0-123.20.1.el7.x86_64 #1
[ 5959.028754] Hardware name: Red Hat RHEV Hypervisor, BIOS 0.5.1 01/01/2007
[ 5959.028756] Workqueue: writeback bdi_writeback_workfn (flush-253:1)
[ 5959.028758]  0000000000000000 00000000c7ca0d4c ffff88004d159900 ffffffff815e2b0c
[ 5959.028761]  ffff88004d159938 ffffffff8105dee1 ffff880029dbeae8 ffff880029dbeae8
[ 5959.028765]  ffff88004d159c50 ffff880029dbe998 ffffea00011383c0 ffff88004d159948
```
http://oss.sgi.com/archives/xfs/2014-07/msg00395.html

looks like another buildbot issue

@dweeezil I'll give your latest patch a try, although I doubt I'll run into any serious lockups due to my (compared to others) minor load

Thanks for your continued work on this issue !

dweeezil commented:
I've pushed a master-rebase of dweeezil/zfs@d85a1f7 (it's dweeezil/zfs@1c05d84).  So far, it's looking very good.  I've not been able to deadlock my test system but I have been able to live-lock it to the point only a SysRq-e will get it going again.

I'm concerned that @tuxoko's observation about the ordering issue with MUTEX_FSTRANS may be happening along the sync path because I'm seeing reclaim entered from ```txg_sync_thread``` in my live-locks:
```
do_try_to_free_pages+0x155/0x440
...
kmem_cache_alloc+0x1b4/0x210
spl_kmem_cache_alloc+0xe9/0xd10 
zio_create+0x84/0x780 
zio_write+0x119/0x1e0 
arc_write+0x121/0x250 
dbuf_write.isra.12+0x24c/0x760 
dbuf_sync_leaf+0x1ac/0xa20 
dbuf_sync_list+0x95/0xe0 
dnode_sync+0x5c5/0x1b60 
dmu_objset_sync_dnodes+0xc5/0x220 
dmu_objset_sync+0x1b8/0x480 
dsl_dataset_sync+0x6a/0x100 
dsl_pool_sync+0xec/0x690 
spa_sync+0x418/0xd50 
txg_sync_thread+0x3e8/0x7f0 
```
I have a feeling this patch's use of ```MUTEX_FSTRANS``` for ```db_mtx``` may be the culprit so I'm going to try  to lock down the individual ```db_mtx``` users and move away from ```MUTEX_FSTRANS``` (and use ```spl_fstrans_mark()``` instead).

chrisrd commented:
@dweeezil  "I've backed off of all the dmu_tx_hold lockdowns. Instead, as @tuxoko rightly pointed out, memory allocations when db_mtx is held can cause deadlocks."

Hmm, that rings a bell...

https://www.illumos.org/issues/5056 - ZFS deadlock on db_mtx and dn_holds

Hasn't been applied to ZoL?

dweeezil commented:
@chrisrd That patch is going to take a bit of digesting; it's pretty extensive.  The stacks shown in https://www.illumos.org/issues/5056 don't look like anything I've seen as part of working on this issue (at least insofar as the umount path is concerned).  That said, I may try a quick port of the patch to see how well it fits into our current code.

behlendorf commented:
> I'm concerned that @tuxoko's observation about the ordering issue with MUTEX_FSTRANS may be happening along the sync path because I'm seeing reclaim entered from txg_sync_thread in my live-locks:

@tuxoko sorry about the long delayed comment on https://github.com/zfsonlinux/zfs/issues/3183#issuecomment-85425528.  You're exactly right, that's a legitimate case which can happen and I should have commented on it explicitly.  At the time I didn't think  that could occur in the existing code but given the stacks observed by @dweeezil it seems it may be possible.  Direct reclaim should never ever ever be possible in the context of the `txg_sync_thread()`.

You mentioned the only way to handle this transparently is to keep a reference count in the task.  Unfortunately, that's not really something we can do because we can't add a field to the task struct.  We could add a bit of thread specific data (TSD) but I suspect the performance hit there would be unacceptable.

We may not be able to transparently hide this case.  Which is a shame because I'd really like to keep the Linux specific bits to a minimum.  I completely agree with @dweeezil's comments above:

> At this rate, the entire system will be reduced to non-sleeping allocations just as it mostly was before the kmem-rework merge. Furthermore, all these semi-gratuitous changes cause even more divergence with upstream code than KM_PUSHPAGE did in the past. I'm going to keep plugging away at it for awhile but a little voice in the back of my head is telling me we need a different approach.

So with that in mind let me throw out a couple possible alternate approaches we could take.  Better ideas are welcome!  I'd love to put this issue behind us once and for all.

* Continue adding targeted calls to `spl_fstrans_mark()` throughout the common ZFS code.  One possible clean way to minimize the footprint here would be to extend the `ZFS_OBJ_HOLD_ENTER/ZFS_OBJ_HOLD_EXIT` macro to take a third `fstrans_cookie_t` argument.  Then we could hide the mark/unmark in the macro for Linux, and a patch could be pushed upstream to add the argument and just ignore it.  I'd have to see how this works out in practice but at the moment this seems like the best approach to me.

* Alternately, we _could_ disable direct reclaim in larger chunks of the code by adding the mark/unmark to the Linux specific `zpl_*` functions.  That would keep the changes in Linux specific code and I think should cover most common cases.  Although it's heavy handed.

* Even more heavy handed would be to just add GFP_NOFS to all KM_SLEEP allocations.  This would be even more comprehensive than the KM_PUSHPAGE solution but it would be straight forward.  We'd definitely want to somehow measure what the performance / stability impact is.  I could see this potentially causing issues on a machine which is starved for memory, but perhaps not.  It's hard to say for sure.

dweeezil commented:
@behlendorf Before addressing your points above, here are a few other details.  First I want to characterize the workload of my torture tests (see https://github.com/dweeezil/stress-tests) to get idea why I'm able to create deadlocks so quickly.  All my testing is done with "mem=4g" to simulate a system that's tight on memory from the beginning.  It's much, much harder to deadlock even at the 8GiB level with all other things equal.  The filesystem is populated with about 1 million files in a 3-level directory hierarchy consisting of 1000 directories.  I've not been re-populating it (right now, there are 899K inodes used on my test system).  A set of 20 concurrent file-creation with random overwrite processes are repeatedly run and at the same time, a set of 20 concurrent random file removal processes are run.  The random removal uses ```find(1)``` to traverse sets of the 1000 directories and randomly removes files.  In other words, I'm running 40 concurrent processes which are rapidly traversing the filesystem.  I wrote these awhile ago to try to create a workload for the issue addressed by bc88866 but starting running into this issue.

This is a mostly metadata-heavy workload and on a memory constrained system, the ARC quickly fills to its automatically-calculated 1GiB cap (and overshoots periodically).

While all this is running, I run a single memory hogging program which nibbles away at a pre-specified amount of memory in 10MiB chunks, dirtying each chunk at it goes.  I'm currently using a 2000MiB cap (somewhat arbitrarily).  The memory hog is allowed to run to completion and is then re-run after a 4-second delay to allow the ARC to refill.

When I launch the memory hog, I watch the ARC via ```arcstats.py``` and, in general, on the first run of the memory hog, I can let it run a good long while without deadlocking.  It will eventually deadlock but my normal regimen is to kill the hogger and let the ARC totally refill and then wait a bit.  The system will generally deadlock on the very next run of the hogger.

With this patch in its current state, I can usually wake the system up by doing a SysRq-e to SIGTERM all the processes.  The trouble seems to stem from the sync task going into reclaim.  I do want to try to back off of using MUTEX_FSTRAMS for ```db_mtx``` and use spl_fstrans_mark as necessary instead.  I do have stack traces saved from a deadlock to help guide me to the trouble spot(s).

As to your points above:

 - I gather that continued use of ```spl_fstrans_mark()``` was the planned path forward and is what I'd like to do to address the deadlocks involving ```db_mtx```.  I do like the idea of encapsulating some of the work in the ```ZFS_OBJ...``` macros to minimize the number of code line additions.  That said, however, similar tricks won't likely work for the potentially many other places in which ```spl_fstrans_mark()``` might be necessary should we continue down that path.
 - I think the scheme of marking the ```zpl_``` functions as well as the sync task might be the way to go for the time being (see below for rationale).
 - I certainly thought of the "always use ```GFP_NOFS```" scheme but it is clearly heavy handed.

My overriding concern is the timing of this issue with respect to a (from what I gather) long-overdue tagged release.  If we miss marking something with ```spl_fstrans_mark()``` by virtue of testing not catching a deadlock, it will most certainly be uncovered once the code gets deployed and if one of these deadlocks is found, it would be big trouble given what appears to be most distros current release model.  This, however, should be mitigated by more widespread adoption of "stable" release updates.

I'd like to add one other note regarding the case of reclaim I spotted in the sync thread:  My first idea was that something was turning off PF_FSTRANS so I put a test and a ```printk()``` at the bottom its loop but the test was never triggered.  In other words, something is apparently disabling PF_FSTRANS and then re-enabling it before hitting the bottom of the sync thread's loop (I did not pepper the loop with any additional tests).

behlendorf commented:
@dweeezil it sounds like we're on the same page about this.  I think the short term goal here needs to be putting together some patch which addresses most/all of the known deadlocks.  Otherwise we're almost certainly going to have people updating to the next released version and encountering this issue.  That's something I definitely want to avoid.

However, this is the last remaining blocker on the list holding up a tag.  So in the interests of getting something out the door, I agree that we either need to adopt the `zpl_` scheme in the short term which I think has a pretty high likelihood of resolving the vast majority of cases.  Or convincing ourselves that your existing patch with a reworked bit for MUTEX_FSTRANS in practice covers most realistic scenarios.  Or, potentially reintroduce a KM_PUSHPAGE replacement for now called something like KM_NOFS.  This would be straight forward to do since 79c76d5b65b19a602d4c7a340da7bf90d4a0c4f8 clearly identifies all the KM_PUSHPAGE site which were changed.  We'd be in no worse shape than the existing released tag and it would be trivial to revert latter.

Have you perhaps already put together a patch for the `zpl_` wrapper case?  If so how do it holding up in testing?  I could propose and test patches for a KM_NOFS approach if your think that's reasonable.

If you could provide a trivial wrapper script in the short term for your test case I could get some additional testing done done myself for these patches.

And speaking of testing thanks for describing your workload and providing a pointer to the test scripts.  It would be ideal if at some point we could some form of your testing to the buildbot, perhaps as part of the zfsstress tests @nedbass put together, https://github.com/nedbass/zfsstress.  These tests are very similar is spirit to the kind of testing you've been doing.

dweeezil commented:
@behlendorf Kind of in reverse order here:  I've put together a wrapper script for my current testing regimen and cleaned up the (very crude) scripts just a bit.  You'll need both https://github.com/dweeezil/stress-tests and https://github.com/dweeezil/junkutils (or your own memory hogging and dirtying program).  The former repository has the stress testing scripts and ```test1.sh``` is the wrapper which will (live|dead)lock my ```mem=4g``` test system very reliably.  The scripts use ```/tank/fish``` as their filesystem and the text in the top of ```test1.sh``` explains how to seed it before testing.  I've not re-seeded my test filesystem in a long time.  One other note, and I suspect it doesn't matter:  I just realized my current test pool has all/most of the newer features disabled (spacemap_histogram, enabled_txg, hole_birth, extensible_dataset, embedded_data and bookmarks).  I've got no idea why but I can't imagine it would matter for this testing.

I've not yet done the ```zpl_...``` wrapper case but it certainly seems it would be pretty easy.  Given that we don't have any upstream with which to maintain compatibility, I don't see any problem with simply peppering each function with the fstrans mark/unmark. I could write a patch this evening and run it through its paces.  Presumably the ZFS_OBJ_HOLD... mark/unmarking I've added could go away as well.

This should give you enough information to get the test case running.  Hopefully you can duplicate the problem as easily as I am able to (which, BTW, I'm strictly running bare metal right now).

dweeezil commented:
@behlendorf I've got a WIP ZPL fstrans mark patch in https://github.com/dweeezil/zfs/tree/zpl-fstrans.  I was still able to deadlock the test system and have observed reclaim during ```zio_execute()``` so I suspect we're running afoul the MUTEX_FSTRANS ordering issue.  With this patch, however, it seems we shouldn't need MUTEX_FSTRANS at all so I'm going to eliminate its use completely.

dweeezil commented:
The ZIO pipeline, although marked as PF_FSTRANS still gets bogged down in reclaim because our shrinker allows it.  The patch in https://github.com/dweeezil/spl/tree/fstrans-shrink-stop (dweeezil/spl@1513b09)  seems to help matters a lot.

It's clear we need something like it to address another problem I've had with the behavior of the system while running these test: a single memory-hogging user program can, effectively, starve the entire ZFS pipeline because we're allowing the system to re-take all our memory.  With this patch in place, the system runs much smoother while running the stress test.

At this point, with the SPL patch, I can no longer deadlock the system but can only live-lock it because the SPL shrinker is returning nonzero counts when reclaim is entered directly from the VFS.  I think we're getting very close.

tuxoko commented:
@dweeezil 
>The ZIO pipeline, although marked as PF_FSTRANS still gets bogged down in reclaim because our shrinker allows it.

Are you sure about this? Don't we turn off __GFP_FS under PF_FSTRANS?
And arc_shrinker_func check for the flags?

dweeezil commented:
@tuxoko I think the problem is that we return a nonzero count, not that the reclaim is actually occurring.  Yes, we turn off the GFP_FS flag during the allocation but the issue here is what happens when the generic shrinker is called by the kernel to get a count and says something _can_ be freed.  It subverts our setting of PF_FSTRANS in zio_execute().  We can't be telling the kernel there's something freeable when we're not in a position to free it.

I've updated the branch referenced by this pull request to the SPL fstrans marking version.

dweeezil commented:
To be clear, this is a problem because we allow some objects to be put on the Linux slab which uses ```kmem_cache_alloc()``` (in the kernel) and _that_ is what calls back into the SPL shrinker.

tuxoko commented:
@dweeezil 
So you're saying that the shrinker functions should return 0 when nr_to_scan is zero(count_objects) and __GFS_FS is not set? It seems reasonable, but the current code doesn't do this.

Also, I don't understand why it would be different for linux slab to call into the shrinker.
We still have wrapper for them to sanitize the flags, have we not?

dweeezil commented:
@tuxoko I wrote my last couple of comments a bit hastily after seeing my system working so much better during the stress test.  The condition I'm able to produce was no longer a deadlock.  The SPL fstrans marking should protect against that.  However, without something to put a stop on attempts to free memory, a single memory-hogging process can effectively cause all I/O to cease making forward progress (the IO pipeline stalls).  So this is not, in fact, a case of "filesystem code" being re-entered during reclaim but it is a case of filesystem _related_ memory attempting to be freed during reclaim.

I'm going to continue running the test today and follow-up with, hopefully, some refinements or further information.

dweeezil commented:
I think I found the elusive (hopefully only remaining) deadlock.  It's a lock inversion between ```shrinker_rwsem``` in the kernel and an unlucky collision in ```z_hold_mtx[]```.  I worked up a patch to annotate each slot in the ```z_hold_mtx[]``` array so I could clearly see collisions.  Details to come shortly once I go over the stack traces.

dweeezil commented:
I updated the patch at https://github.com/dweeezil/spl/tree/fstrans-shrink-stop to be a lot more forceful and simply don't allow any shrinking at all when PF_FSTRANS is in effect.  When running the stress test, however, all read operations as seen by ```iostat``` will sometimes hang for long periods of time.  Interestingly enough, it always seems to recover.

The lock inversion I thought I found in an earlier comment was a false alarm so I deleted the comment (from github).

kpande commented:
I was doing some benchmarking last night after introducing large block patches to the latest master and I noticed I can't actually complete a certain 'fio' benchmark:
```
    [global]  
    description=Emulation of Intel IOmeter File Server Access Pattern  
      
    [iometer]  
    bssplit=8k/100  
    rw=randrw  
    rwmixread=50  
    direct=0
    size=44g  
    ioengine=libaio  
    # IOMeter defines the server loads as the following:  
    # iodepth=1     Linear  
    # iodepth=4     Very Light  
    # iodepth=8     Light  
    # iodepth=64    Moderate  
    # iodepth=256   Heavy  
    iodepth=32  
```

The original version of this test used DirectIO so I had to disable it here to launch directly on ZPL.

My experience was:

* Testing via iSCSI client, booted from zvol with small max blocksize (8k), achieved 75MB/s both ways, about 9700IOPS

* Testing locally on ZPL or ZVOL would give me 3MB/s with ZPL causing load increasing on server up to 60+ to the point that the test will not complete. zvol do not cause this kind of load spike.

I shut down the benchmark and the system was still feeding data to disks for about 5 minutes before the IO relaxed and was at / near zero again.

* Tested via iSCSI client, same zvol, no changes (didn't even clear previous test file) and it does about 3MB/s or ~700 IOPS. Huge decrease.

Set primarycache=none to the iSCSI image and noted that it began at 75MB/s and dropped to 23MB/s or ~1500 IOPS.

behlendorf commented:
@dweeezil this looking good, I like it as a solution much better than I thought I would.  Setting PF_FSTRANS in the VFS callbacks is clean and has the nice side-effect of self-documenting the code.  Basically, while in a VFS ZPL callback don't allow reentry in to the VFS.

I'm happy to put some additional test mileage on these patches, just let me know when they're in a state you're happy with.

kernelOfTruth commented:
referencing https://github.com/zfsonlinux/zfs/pull/2828 here if any changes related to the ZIO pipeline needs to be made and if this might help

dweeezil commented:
@behlendorf Please give it some testing.  I'll not have a chance to exercise it more until later this evening.  Also, we're likely going to need some of the SPL patch I've referenced above.

tuxoko commented:
@dweeezil 
I was testing my direct reclaim torture test on your patch and I get this
```
[ 2640.136202] INFO: task create.sh:22272 blocked for more than 120 seconds.
[ 2640.136205]       Tainted: P           OE  3.18.10 #7
[ 2640.136206] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 2640.136208] create.sh       D ffff880118f162a0     0 22272  14798 0x00000000
[ 2640.136211]  ffff88003bc3b7b8 0000000000000086 ffff880118f15d60 ffff88003bc3bfd8
[ 2640.136214]  00000000000135c0 00000000000135c0 ffff880119bc94c0 ffff880118f15d60
[ 2640.136217]  ffff88003bc3b798 ffff880038a6c9a8 ffff880038a6c9a8 ffff880038a6c9b0
[ 2640.136219] Call Trace:
[ 2640.136223]  [<ffffffff8175c831>] schedule_preempt_disabled+0x31/0x80
[ 2640.136227]  [<ffffffff8175f913>] __mutex_lock_slowpath+0x123/0x440
[ 2640.136231]  [<ffffffff8175fc61>] mutex_lock+0x31/0x45
[ 2640.136250]  [<ffffffffc076843d>] txg_hold_open+0x4d/0x140 [zfs]
[ 2640.136262]  [<ffffffffc0721c97>] dmu_tx_assign+0x277/0x5e0 [zfs]
[ 2640.136274]  [<ffffffffc0722d79>] ? dmu_tx_hold_sa+0x179/0x1a0 [zfs]
[ 2640.136293]  [<ffffffffc07a5af3>] zfs_inactive+0x173/0x290 [zfs]
[ 2640.136312]  [<ffffffffc07be467>] zpl_evict_inode+0x47/0x60 [zfs]
[ 2640.136317]  [<ffffffff811fbb78>] evict+0xb8/0x190
[ 2640.136321]  [<ffffffff811fbc91>] dispose_list+0x41/0x50
[ 2640.136325]  [<ffffffff811fcc87>] prune_icache_sb+0x47/0x60
[ 2640.136329]  [<ffffffff811e3985>] super_cache_scan+0x105/0x170
[ 2640.136347]  [<ffffffffc07be620>] ? zpl_show_options+0x90/0x90 [zfs]
[ 2640.136365]  [<ffffffffc079b088>] zfs_sb_prune+0x48/0x60 [zfs]
[ 2640.136382]  [<ffffffffc07be646>] zpl_prune_sb+0x26/0x30 [zfs]
[ 2640.136390]  [<ffffffffc06f9573>] arc_do_user_prune+0xb3/0x230 [zfs]
[ 2640.136399]  [<ffffffffc06fb289>] zfs_debug_direct_reclaim_func+0x69/0x90 [zfs]
[ 2640.136404]  [<ffffffffc0650313>] spl_kmem_debug_direct_reclaim+0x43/0x50 [spl]
[ 2640.136417]  [<ffffffffc0737d4c>] dsl_dir_tempreserve_space+0x5c/0x1a0 [zfs]
[ 2640.136428]  [<ffffffffc0721f9a>] dmu_tx_assign+0x57a/0x5e0 [zfs]
[ 2640.136445]  [<ffffffffc07a3954>] zfs_dirty_inode+0xd4/0x350 [zfs]
[ 2640.136449]  [<ffffffff81757003>] ? __slab_free+0x101/0x2c2
[ 2640.136453]  [<ffffffff81757003>] ? __slab_free+0x101/0x2c2
[ 2640.136498]  [<ffffffffc07be48e>] zpl_dirty_inode+0xe/0x10 [zfs]
[ 2640.136503]  [<ffffffff8120c9a2>] __mark_inode_dirty+0xc2/0x2c0
[ 2640.136520]  [<ffffffffc07a972d>] zfs_mark_inode_dirty+0x4d/0x60 [zfs]
[ 2640.136538]  [<ffffffffc07bc68a>] zpl_release+0x3a/0xa0 [zfs]
[ 2640.136542]  [<ffffffff811e1cb7>] __fput+0xf7/0x230
[ 2640.136545]  [<ffffffff811e1e3e>] ____fput+0xe/0x10
[ 2640.136549]  [<ffffffff8108e2f4>] task_work_run+0xb4/0xe0
[ 2640.136554]  [<ffffffff8101409f>] do_notify_resume+0x8f/0xb0
[ 2640.136558]  [<ffffffff81761f6f>] int_signal+0x12/0x17
```
So maybe FSTRANS should cover zfs_mark_inode_dirty

dweeezil commented:
@tuxoko Thanks, good catch.  Update pushed as dweeezil/zfs@714a5c9.

dweeezil commented:
Something I've been wanting to bring up in this issue is the raising of ```ZFS_OBJ_MTX_SZ``` to 256.  Granted, this isn't a big deal as far as memory use is concerned but in retrospect  I think most of the rationale referenced in 30315d2 has been long since superseded by other work culminating with the kmem-rework merge. @behlendorf: Mentioning you here because it was your commit.

As I alluded to earlier in this thread, as part of my work on making ZoL compatible with ```CONFIG_PROVE_LOCKING``` and to, in general, to make it more lockdep friendly, I have come up with a couple of patches which, among other things, fan-out the entire ```z_hold_mtx[]``` array so each element of each filesystem is in its own "mutex class" and, therefore, shows up uniquely in lockdep output.  This has been invaluable in showing the effects of hash collisions in the array.

What I've not mentioned yet, however, is that I've not seen a whole lot of actual contention on this array of locks so I'd like to float the idea of reverting it back to 64 elements as appears to be used by all other OpenZFS implementations.  My feeling is that raising it to 256 was more of a band-aid and that the outcome of this current exercise should eliminate the "benefit" of raising it.

Thoughts... ?

tuxoko commented:
@dweeezil 
I would like to hear about what so far you have done for CONFIG_PROVE_LOCKING. I also tried to use it once, and the most annoying nesting lock for me is db_mtx. I tried to solve that by applying different different class to different DMU object type but to no avail because there are interdependence on dsl_dataset.

tuxoko commented:
My direct reclaim torture mechanism works better than I expected.
I found another deadlock on z_teardown_inactive_lock:
```
[ 4680.122832] INFO: task zfs:5064 blocked for more than 120 seconds.
[ 4680.122843]       Tainted: P           OE  3.18.10 #7
[ 4680.122852] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 4680.122862] zfs             D ffff8800b3a18fa0     0  5064   4863 0x00000000
[ 4680.122876]  ffff880092083630 0000000000000086 ffff8800b3a18a60 ffff880092083fd8
[ 4680.122890]  00000000000135c0 00000000000135c0 ffff880119bc8a60 ffff8800b3a18a60
[ 4680.122903]  0000000000000000 ffff8800b3a18a60 ffff8800c73f6d40 ffff8800c73f6d58
[ 4680.122917] Call Trace:
[ 4680.122937]  [<ffffffff8175c309>] schedule+0x29/0x70
[ 4680.122958]  [<ffffffff81760475>] rwsem_down_read_failed+0xc5/0x120
[ 4680.122983]  [<ffffffff8139cd94>] call_rwsem_down_read_failed+0x14/0x30
[ 4680.123004]  [<ffffffff8175fe47>] ? down_read+0x37/0x40
[ 4680.123106]  [<ffffffffc07ec890>] zfs_inactive+0x40/0x290 [zfs]
[ 4680.123208]  [<ffffffffc0805327>] zpl_evict_inode+0x47/0x60 [zfs]
[ 4680.123233]  [<ffffffff811fbb78>] evict+0xb8/0x190
[ 4680.123254]  [<ffffffff811fbc91>] dispose_list+0x41/0x50
[ 4680.123302]  [<ffffffff811fcc87>] prune_icache_sb+0x47/0x60
[ 4680.123324]  [<ffffffff811e3985>] super_cache_scan+0x105/0x170
[ 4680.123419]  [<ffffffffc08054e0>] ? zpl_show_options+0x90/0x90 [zfs]
[ 4680.123509]  [<ffffffffc07e1f58>] zfs_sb_prune+0x48/0x60 [zfs]
[ 4680.123597]  [<ffffffffc0805506>] zpl_prune_sb+0x26/0x30 [zfs]
[ 4680.123637]  [<ffffffffc0740573>] arc_do_user_prune+0xb3/0x230 [zfs]
[ 4680.123678]  [<ffffffffc0742289>] zfs_debug_direct_reclaim_func+0x69/0x90 [zfs]
[ 4680.123704]  [<ffffffffc06aa313>] spl_kmem_debug_direct_reclaim+0x43/0x50 [spl]
[ 4680.123776]  [<ffffffffc0752808>] dbuf_hold_impl+0x38/0xb0 [zfs]
[ 4680.123823]  [<ffffffffc0752bb0>] dbuf_hold+0x20/0x40 [zfs]
[ 4680.123874]  [<ffffffffc075bcff>] dmu_buf_hold_noread+0x6f/0x170 [zfs]
[ 4680.123923]  [<ffffffffc075be2f>] dmu_buf_hold+0x2f/0x80 [zfs]
[ 4680.123944]  [<ffffffff814a52f0>] ? extract_entropy+0xc0/0x150
[ 4680.124010]  [<ffffffffc07c5e41>] zap_lockdir+0x61/0x970 [zfs]
[ 4680.124188]  [<ffffffff8139a851>] ? vsnprintf+0x181/0x590
[ 4680.124232]  [<ffffffffc0742246>] ? zfs_debug_direct_reclaim_func+0x26/0x90 [zfs]
[ 4680.124316]  [<ffffffffc07c6e5f>] zap_lookup_norm+0x3f/0x1b0 [zfs]
[ 4680.124400]  [<ffffffffc07c7003>] zap_lookup+0x33/0x40 [zfs]
[ 4680.124466]  [<ffffffffc07825ac>] dsl_prop_get_dd+0xfc/0x210 [zfs]
[ 4680.124531]  [<ffffffffc0782836>] dsl_prop_get_ds+0x176/0x230 [zfs]
[ 4680.124563]  [<ffffffffc06b02f7>] ? tsd_hash_search.isra.1+0x77/0xa0 [spl]
[ 4680.124588]  [<ffffffffc06a7a97>] ? spl_kmem_alloc+0xf7/0x1e0 [spl]
[ 4680.124651]  [<ffffffffc078294d>] dsl_prop_register+0x3d/0x1b0 [zfs]
[ 4680.124771]  [<ffffffffc07e1d90>] ? zfs_is_readonly+0x20/0x20 [zfs]
[ 4680.124859]  [<ffffffffc07e20f4>] zfs_register_callbacks+0x74/0x4a0 [zfs]
[ 4680.124945]  [<ffffffffc07e38d7>] zfs_sb_setup+0x17/0x1e0 [zfs]
[ 4680.125030]  [<ffffffffc07e4942>] zfs_resume_fs+0x242/0x420 [zfs]
[ 4680.125141]  [<ffffffffc07d997e>] zfs_ioc_rollback+0x8e/0xa0 [zfs]
[ 4680.125168]  [<ffffffffc068e8ea>] ? nvlist_alloc+0x4a/0x50 [znvpair]
[ 4680.125252]  [<ffffffffc07d9be5>] zfsdev_ioctl+0x1d5/0x480 [zfs]
[ 4680.125423]  [<ffffffff811f3890>] do_vfs_ioctl+0x300/0x520
[ 4680.125443]  [<ffffffff811f3b31>] SyS_ioctl+0x81/0xa0
[ 4680.125465]  [<ffffffff81761ccd>] system_call_fastpath+0x16/0x1b
```

tuxoko commented:
@dweeezil 
Could you cherry pick this https://github.com/tuxoko/zfs/commit/6ef68717c20202b5f48b3a5b0888dc8832a3288d
This revert the MUTEX_FSTRANS on l2arc_buflist_mtx and use spl_fstrans_mark on l2arc_feed_thread.
Thanks

dweeezil commented:
@tuxoko I've done the cleanup you mentioned and also added you patch for the l2arc feed thread.  I had been meaning to look at those as well.  Do I gather the arc adapt thread doesn't need PF_FSTRANS?

My current stress test was written for something completely different and does not exercise any of the management ioctls.  Regarding your trace from ```zfs_ioc_rollback()```, in the same vein as we've wrapped all the ZPL functions, I suppose we ought to wrap _all_ the ioctls.  I've done just that in the just-pushed commit.

Regarding my earlier attempts at using CONFIG_PROVE_LOCKING, I actually gave up on it and planned on re-visiting it once this issue was handled.

dweeezil commented:
@tuxoko I see you've got an SPL cleanup patch in the works.  I'll remove the definition of MUTEX_FSTRANS on the ZFS side as part of this one.

tuxoko commented:
@dweeezil 
It seems arc_adapt does need PF_FSTRANS.
By the way, I just push out my direct reclaim torture test:
https://github.com/zfsonlinux/spl/pull/442
https://github.com/zfsonlinux/zfs/pull/3246
You can check them out if you're interested.

I catched quite some deadlocks with it. I also got a lockup/panic once, but I didn't set up the serial console so I'm not sure what happened.

dweeezil commented:
@tuxoko Thanks, I was going to ask about your direct reclaim torture test.  I'll get arc_adapt locked down shortly.  It doesn't surprise me at all that it needs it.

behlendorf commented:
My only concern regarding decreasing `ZFS_OBJ_MTX_SZ` back to 64 is that it will decrease the number of concurrent lookups.  For most workloads I doubt this would be an issue, and I doubt you'd see contention because you're not going to have a huge number of threads.  However, for workloads like a Lustre meta data server it's not uncommon for there to be 1000's of threads which are all constantly doing lookups.  So before we go decreasing this value we should get some performance results from a workload, like Lustre, which does this and see how significant decreasing that value would be.

I'll setup my own machine today with the latest patches to stress test this code too.

tuxoko commented:
Hi @behlendorf 
I got this kernel panic when doing torture testing.
```
[ 4692.202686] Kernel panic - not syncing: thread ffff8800c4f5dd60 terminating with rrw lock ffff8800da1b9c40 held
[ 4692.228053] CPU: 1 PID: 6250 Comm: mmap_deadlock Tainted: P           OE  3.18.10 #7
```
Unfortunately, it didn't show any stack trace, but by looking into the panic message, the culprit should be ```tsd_exit()``` in ```ZFS_EXIT(zsb)```
The comment for says ```tsd_exit - destroys all thread specific data for this thread```.

I don't think it would be wise to destroy ALL tsd data just for exit a certain rrwlock, would it?
The right way to do this should be check the remaining entries during rrw_exit, and only destroy rrw_tsd_key if there are none left.

behlendorf commented:
> I don't think it would be wise to destroy ALL tsd data just for exit a certain rrwlock, would it?

@tuxoko I can see how the comment here could be slightly misleading.  We're destroying all the tsd data which which is being tracked by the `tsd_` interfaces which are being emulated by the SPL.  Just to be clear, user tsd is managed entirely separately and is unrelated.

Ideally the _right_ time to do this would be when the thread is destroyed.  Unfortunately, the kernel doesn't provide us any mechanism we can use to be notified that a thread has been destroyed.  That means to avoid leaking tsd data we need to make sure it get's destroyed when we exit the VFS hook because ZFS may never see that thread again.  This means the tsd data is only valid when the task in the ZFS code but as it turns out that's good enough for our purposes.

What would be ideal is to remove our need to tsd data entirely since the kernel doesn't provide us a tsd  interface.  It's used very infrequently in ZFS so I'm sure it's doable but it would take some effort.  This is really an entirely separate issue which we should file a new issue for.

behlendorf commented:
@dweeezil @tuxoko thus far my overnight testing has held up well with the latest version of this patch.  I've been running @dweeezil's torture tests and aside from one unrelated panic, and a small fix to the spl patch, I haven't observed any problems.  This is definitely a significant improvement.  How is it holding up in your testing?

dweeezil commented:
@behlendorf So far, this is looking much better, however, on my stress test suite, I'm still seeing all I/O come to a complete halt after awhile (last time it took 100 hog iterations which works out to probably 25 minutes of run-time.  Here's a couple of other notes regarding my test setup.
    - I'm using a single partition on a pretty fast SSD for my pool, so there's not a whole lot of resistance to the I/O requests.
    - I run an instance of "arcstats.py 1" and "iostat 1" in their own windows (this is all running using the VGA console).
    - In my last test run, I stressed the system further by tarring a linux kernel git repo on a the ext4 root filesystem.

While I've been typing this, my current run has stalled I/O after 65 hog iterations.  This time I'm going to capture a SysRq-t (I'd been mainly using SysRq-w previously).  This I/O hang is _not_ a deadlock;  I can un-wedge the system with a SysRq-i.  I just got the 120-second warnings on the console so I'm going to go and take my SysRq-t (and a stand-alone SysRq-d for good measure) and the start going over the stacks.

dweeezil commented:
Just as I was about to go and get some stack traces, the I/O started again on my test system.  It appears to have been stalled for at least several minutes (long enough for a bunch of tasks to trigger the 120-second warning).   I'm going to wait for it to lock up again and get the SysRq-t stacks.

One other thing, @behlendorf, are you running with the SPL patch?  Without it, things are much much worse.  In particular, interactive operation in a free console window is practically unusable.

behlendorf commented:
@dweeezil my system just entered a similar state after 38 iterations of your torture test.  I was able to grab the stack traces.  In my case the cause appears to be that the `z_wr_iss` threads are having a very hard time allocating memory for the compressed IO buffer in `zio_write_bp_init()->zio_buf_alloc()`.  They are making forward progress so this _isn't_ a deadlock and given enough time I expect they'll complete.  In the meanwhile the system in unresponsive.

Given the stress we're putting the system under this doesn't surprise me too much.  It's also something I expect @tuxoko's ABD patches will help significantly with when they're merged.  I fully expect the existing 0.6.3 code  would behave the same way given the conditions.

I have been using a slightly modified version of the SPL patch and it has helped considerably.  I'm using the following patch.

```diff
diff --git a/module/spl/spl-kmem-cache.c b/module/spl/spl-kmem-cache.c
index 6fcc7c4..cd3e543 100644
--- a/module/spl/spl-kmem-cache.c
+++ b/module/spl/spl-kmem-cache.c
@@ -1574,6 +1574,12 @@ __spl_kmem_cache_generic_shrinker(struct shrinker *shrink
        spl_kmem_cache_t *skc;
        int alloc = 0;
 
+       /*
+        * No shrinking in a transaction context.  Can cause deadlocks.
+        */
+       if (sc->nr_to_scan && spl_fstrans_check())
+               return (SHRINK_STOP);
+
        down_read(&spl_kmem_cache_sem);
        list_for_each_entry(skc, &spl_kmem_cache_list, skc_list) {
                if (sc->nr_to_scan) {
```

dweeezil commented:
@behlendorf You've pretty much confirmed exactly what I'm seeing.  I mainly wanted to look at _all_ the stacks because frequently the issue threads aren't dormant so they don't show up in SysRq-w.  I had been meaning to try some older code but didn't feel like messing around with it to get it compiled on the newer kernels on which I'm testing.  I fully expect this test would crush a system running stock 0.6.3.

I just updated the SPL patch to use the helper function and will post it as a pull request shortly.

dweeezil commented:
@behlendorf What are your thoughts on trying to tune ```zio_taskqs[]``` on small memory and/or low core-count systems?  I realize it's a completely different issue than the original one in this thread but it seems like it might be helpful in some cases (Raspberry Pi and other small-memory systems).

behlendorf commented:
@dweeezil when you refresh the SPL patch can you also revert the MUTEX_FSTRANS  patch zfsonlinux/spl@d0d5dd714424365a4da0d887cb641cb2f0ae8844.  Let's get rid of that interface since we've decided not to use it.

As for the `zio_taskqs[]` I've thought for a while we should at a minimum make those values tunable somehow.  That would allow us to more easily get feedback on reasonable values for a variety of systems.

behlendorf commented:
@dweeezil @tuxoko  the latest version of the patch stack continues to hold together well for me in all but the most abusive testing.  In my estimation it is behaving at least as well as, if not better than, the existing 0.6.3 code when in a low memory situation.  There's certainly room for us to continue improving things, and we should, but in the interests of getting this our due tag I'd like to merge this.  Then we can get a final round of testing in over the weekend before the tag.

Do you guys have any reservations or last minute improvements you'd like to get in to this change?  What's your general impression of the state of things with this patch applied to master?

dweeezil commented:
@behlendorf Insofar as fixing the original issues described in this thread and some other related issues, I'm happy with this patch as it stands.  What's your take on the SPL patch?  I've been doing all my testing with it as it helps things a lot in this particular case.

My impression is that things are in pretty good shape with this patch applied.  I'm planning doing some more tests this weekend on bigger hardware configurations along with my (delayed by this issue) testing of pull request 3115 (purposely not hash-referenced; It was this issue which stopped me in my tracks while working on that patch).

behlendorf commented:
@dweeezil I've been using the SPL patch as well and I think it helps considerably in certain situations.  I was going to include it when I merge this change.  Thus far it's worked well for me after making that one minor fix.  OK, then I'll get this merged today and queue up some additional weekend testing on the new master.

deajan commented:
Hi,
Since i upgraded my until now very stable backup server from CentOS 7.0 to 7.1, i had to upgrade zfs-testing from zfs-0.6.3-166 to zfs-0.6.3-260.
I now have a lot of deadlocks like this one, which end up totally locking my machine, from ssh and console.

```
Apr  5 05:53:22 backupmaster kernel: INFO: task kswapd0:46 blocked for more than 120 seconds.
Apr  5 05:53:22 backupmaster kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Apr  5 05:53:22 backupmaster kernel: kswapd0         D ffff88021fc13680     0    46      2 0x00000000
Apr  5 05:53:22 backupmaster kernel: ffff880210fab778 0000000000000046 ffff880210fabfd8 0000000000013680
Apr  5 05:53:22 backupmaster kernel: ffff880210fabfd8 0000000000013680 ffff880213d64440 ffff880213d64440
Apr  5 05:53:22 backupmaster kernel: ffff8800d22ee570 fffffffeffffffff ffff8800d22ee578 0000000000000001
Apr  5 05:53:22 backupmaster kernel: Call Trace:
Apr  5 05:53:22 backupmaster kernel: [<ffffffff81609e29>] schedule+0x29/0x70
Apr  5 05:53:22 backupmaster kernel: [<ffffffff8160b925>] rwsem_down_read_failed+0xf5/0x165
Apr  5 05:53:22 backupmaster kernel: [<ffffffff812ede0d>] ? list_del+0xd/0x30
Apr  5 05:53:22 backupmaster kernel: [<ffffffff812e31e4>] call_rwsem_down_read_failed+0x14/0x30
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa02b59d2>] ? spl_kmem_free+0x32/0x50 [spl]
Apr  5 05:53:22 backupmaster kernel: [<ffffffff816091f0>] ? down_read+0x20/0x30
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03effeb>] dmu_zfetch_find+0x4b/0x830 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffff812ede0d>] ? list_del+0xd/0x30
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03f0add>] dmu_zfetch+0xcd/0x930 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03d7249>] dbuf_read+0xa99/0xbd0 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03f2548>] dnode_hold_impl+0x198/0x610 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811ae07d>] ? __kmalloc_node+0x13d/0x280
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03f29d9>] dnode_hold+0x19/0x20 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03ed732>] dmu_tx_hold_object_impl.isra.1+0x42/0x190 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03ed895>] dmu_tx_hold_bonus+0x15/0x30 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa03ef01c>] dmu_tx_hold_sa+0x3c/0x180 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa047302b>] zfs_inactive+0x15b/0x290 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffffa048b783>] zpl_evict_inode+0x43/0x60 [zfs]
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811e2237>] evict+0xa7/0x170
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811e233e>] dispose_list+0x3e/0x50
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811e3213>] prune_icache_sb+0x163/0x320
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811ca496>] prune_super+0xd6/0x1a0
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811691a5>] shrink_slab+0x165/0x300
Apr  5 05:53:22 backupmaster kernel: [<ffffffff811c0751>] ? vmpressure+0x21/0x90
Apr  5 05:53:22 backupmaster kernel: [<ffffffff8116cdf1>] balance_pgdat+0x4b1/0x5e0
Apr  5 05:53:22 backupmaster kernel: [<ffffffff8116d093>] kswapd+0x173/0x450
Apr  5 05:53:22 backupmaster kernel: [<ffffffff81098340>] ? wake_up_bit+0x30/0x30
Apr  5 05:53:22 backupmaster kernel: [<ffffffff8116cf20>] ? balance_pgdat+0x5e0/0x5e0
Apr  5 05:53:22 backupmaster kernel: [<ffffffff8109738f>] kthread+0xcf/0xe0
Apr  5 05:53:22 backupmaster kernel: [<ffffffff810972c0>] ? kthread_create_on_node+0x140/0x140
Apr  5 05:53:22 backupmaster kernel: [<ffffffff8161497c>] ret_from_fork+0x7c/0xb0
Apr  5 05:53:22 backupmaster kernel: [<ffffffff810972c0>] ? kthread_create_on_node+0x140/0x140
```

I am currently connected to my machine, which is quite unresponsive, top showing a load average of 18, but no high cpu usage nor disk usage (wa = 0.9).

I've setup a crontab that logs free memory and arcstats every 3 minutes until it will crash.

Current stats are
```
[root@backupmaster ~]# free -h
              total        used        free      shared  buff/cache   available
Mem:           7,6G        4,2G        343M        5,3M        3,1G        1,4G
Swap:          7,9G        272M        7,6G
[root@backupmaster ~]# cat /proc/spl/kstat/zfs/arcstats
5 1 0x01 86 4128 10440723686 103106010736704
name                            type data
hits                            4    6799000
misses                          4    903734
demand_data_hits                4    1052478
demand_data_misses              4    98478
demand_metadata_hits            4    3979439
demand_metadata_misses          4    308492
prefetch_data_hits              4    90300
prefetch_data_misses            4    389395
prefetch_metadata_hits          4    1676783
prefetch_metadata_misses        4    107369
mru_hits                        4    2804169
mru_ghost_hits                  4    68374
mfu_hits                        4    2227748
mfu_ghost_hits                  4    17802
deleted                         4    1092616
recycle_miss                    4    95086
mutex_miss                      4    95
evict_skip                      4    34452369
evict_l2_cached                 4    0
evict_l2_eligible               4    94026845184
evict_l2_ineligible             4    3055902720
hash_elements                   4    37261
hash_elements_max               4    121953
hash_collisions                 4    95936
hash_chains                     4    723
hash_chain_max                  4    4
p                               4    1261568616
c                               4    1261827688
c_min                           4    4194304
c_max                           4    4085360640
size                            4    1261617624
hdr_size                        4    13965568
data_size                       4    52940800
meta_size                       4    311249920
other_size                      4    883461336
anon_size                       4    3293184
anon_evict_data                 4    0
anon_evict_metadata             4    0
mru_size                        4    341787136
mru_evict_data                  4    40685568
mru_evict_metadata              4    2132480
mru_ghost_size                  4    57884160
mru_ghost_evict_data            4    5898240
mru_ghost_evict_metadata        4    51985920
mfu_size                        4    19110400
mfu_evict_data                  4    12255232
mfu_evict_metadata              4    72192
mfu_ghost_size                  4    250805248
mfu_ghost_evict_data            4    76679168
mfu_ghost_evict_metadata        4    174126080
l2_hits                         4    0
l2_misses                       4    0
l2_feeds                        4    0
l2_rw_clash                     4    0
l2_read_bytes                   4    0
l2_write_bytes                  4    0
l2_writes_sent                  4    0
l2_writes_done                  4    0
l2_writes_error                 4    0
l2_writes_hdr_miss              4    0
l2_evict_lock_retry             4    0
l2_evict_reading                4    0
l2_free_on_write                4    0
l2_cdata_free_on_write          4    0
l2_abort_lowmem                 4    0
l2_cksum_bad                    4    0
l2_io_error                     4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
l2_compress_successes           4    0
l2_compress_zeros               4    0
l2_compress_failures            4    0
memory_throttle_count           4    0
duplicate_buffers               4    0
duplicate_buffers_size          4    0
duplicate_reads                 4    0
memory_direct_count             4    1361
memory_indirect_count           4    91273
arc_no_grow                     4    0
arc_tempreserve                 4    0
arc_loaned_bytes                4    0
arc_prune                       4    0
arc_meta_used                   4    1208676824
arc_meta_limit                  4    3064020480
arc_meta_max                    4    1220892016
```

I get more and more load average and will probably loose "the hand" on the machine.
Anything i could try ? Anything i should log to help finding where the deadlock comes from ?

Running CentOS 7.1 Linux backupmaster.siege.local 3.10.0-229.1.2.el7.x86_64 #1 SMP Fri Mar 27 03:04:26 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux on a Fujitsu TX140S1P with 2x4Tb RE4 drives and 8Gb RAM, no deduplication.

PS: sorry if this isn't the right issue to post to, i just read a lot of deadlock stuff and this one just looks like mine.

DeHackEd commented:
There should be more than one stack trace. Can you paste (or pastebin, or gist, etc) some more from the same time period? Should all be logged within a 2 minute window.

Also note that another commit went in just recently to improve the situation further. It's not a 100% fix but every little bit helps. Pulling the next update would be my suggestion.

deajan commented:
Well there is a zfs-0.6.3-261 package, but it seems it isn't public or built yet (so it was yesterday).
It says HTTP 404 for the spl and zfs rpms when i try yum --enablerepo=zfs-testing update

The full stack trace can be found here: https://gist.github.com/deajan/78757af1066b7d77a73d

behlendorf commented:
@deajan Sorry about that.  The EPEL zfs-testing repository will be fixed today.

deajan commented:
@behlendorf Thanks a lot, and congrats for release 0.6.4 :))

ryao commented:
It would be better to identify each lock, convert the entry/exit routines into preprocessor macro wrappers and insert `spl_fstrans_mark()`/`spl_fstrans_unmark()` there. That way it would be clear why a given location needs it when doing review in the future. Doing it in the VFS hooks will miss this class of problems in code paths exercised by taskq threads.
